### PR TITLE
Refactor Pulley's interpreter loop

### PR DESCRIPTION
**Pull request by alexcrichton**
* Define loop-over-match and loop-with-tail-calls in separate files to make it more clear which is in which (and less `#[cfg]`)
* Move per-opcode handlers to `interp.rs` outside of a macro invocation to get better native editor support (e.g. formatting, hints, etc).

This is roughly intended to be perf-neutral but we don't have many automated benchmarks yet for Pulley so it's intended to profile later as well.

### CODE REVIEW

**Comment by abrown** [comment]
@fitzgen, I'll leave this for you for when you get back as I assume you have stronger opinions on this level of changes.

**Comment by Kmeakin** [comment]
> cc @fitzgen @Kmeakin 

No objections from me

**Comment by alexcrichton** [comment]
I've done a bit more refactoring here with the goal of modeling the interpreter as an implementation of `OpVisitor` and `ExtendedOpVisitor`. This should help improve error messages and type-checking for missing operations when new ones are added.

**Review by abrown — APPROVED** [review]
Makes sense to me; take my comment below with a grain of salt since I wasn't involved in initially designing this (but, yeah, it isn't too clear _when_ interpretation is occurring in the two modes).

File: pulley/src/interp/match_loop.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+use super::*;
+use crate::decode::unwrap_uninhabited;
+
+impl Interpreter<'_> {
+    pub fn run(mut self) -> Done {
+        let mut decoder = Decoder::new();
+        loop {
+            match unwrap_uninhabited(decoder.decode_one(&mut self)) {
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply pulley/src/interp/match_loop.rs:32]: Good points! I've tried to add more words here to explain what's going on

**Comment by alexcrichton** [comment]
Ok I'm going to flag this for merging but @fitzgen you'll definitely be interested in this for when you get back, and I'm happy to address any feedback you have in follow-ups.

**Review by fitzgen — COMMENTED** [review]
LGTM!

**Pull request merged**